tools/shared/src/main/scala/org/scalajs/core/tools/sourcemap/SourceMapWriter.scala

```diff
@@ -50,8 +53,48 @@ class SourceMapWriter(
   private val _nameToIndex = new HashMap[String, Int]
 
   // Strings are nullable in this stack
-  private val nodePosStack = new Stack[(Position, String)]
-  nodePosStack.push((NoPosition, null))
+  private val nodePosStack = new NodePosStack
+  nodePosStack.push(NoPosition, null)
+
+  private final class NodePosStack {
+    private var topIndex: Int = -1
+    private var posStack: Array[Position] = new Array(32)
+    private var nameStack: Array[String] = new Array(32)
+
+    def pop(): Unit = {
+      val topIdx = topIndex
+      val newTopIdx = topIdx - 1
+      topIndex = newTopIdx
+      posStack(topIdx) = null
+      nameStack(topIdx) = null
+    }
+
+    def topPos: Position =
+      posStack(topIndex)
+
+    def topName: String =
+      nameStack(topIndex)
+
+    def push(pos: Position, originalName: String): Unit = {
+      val newTopIdx = topIndex + 1
+      topIndex = newTopIdx
+      val currentSize = posStack.length
+      if (newTopIdx >= currentSize)
+        changeStackSize(2 * currentSize, newTopIdx)
+      posStack(newTopIdx) = pos
+      nameStack(newTopIdx) = originalName
+
+    }
+
+    private def changeStackSize(newSize: Int, copyLength: Int): Unit = {
+      val newPosStack = new Array[Position](newSize)
+      val newNameStack = new Array[String](newSize)
+      System.arraycopy(posStack, 0, newPosStack, 0, copyLength)
+      System.arraycopy(nameStack, 0, newNameStack, 0, copyLength)
+      posStack = newPosStack
+      nameStack = newNameStack
```

Comment:
Consider using `java.util.Arrays.copy(array, newLength)`, which basically does all of that.